### PR TITLE
Ignore strict loading violations on instances loaded through fixtures

### PR DESCRIPTION
**Pull request by ghiculescu**
Fixes an issue reported here: https://discuss.rubyonrails.org/t/how-to-eager-load-fixture-associations-in-rails-6-1/76645

### CODE REVIEW

File: activerecord/test/cases/fixtures_test.rb [hunk 1]
```diff
@@ -1589,3 +1591,18 @@ def readonly_config
...
 end
+
+class StrictLoadingFixturesTest < ActiveRecord::TestCase
+  fixtures :strict_zines
+
+  test "strict loading violations are ignored on fixtures" do
+    ActiveRecord::FixtureSet.reset_cache
+    create_fixtures("strict_zines")
+
+    assert_raises(ActiveRecord::StrictLoadingViolationError) do
```
Comment (ghiculescu) [inline activerecord/test/cases/fixtures_test.rb:1602]: this assertion is just to show that strict loading is enabled on this model. by contrast, the next line doesn't crash.

**Comment by ghiculescu** [comment]
@eileencodes just wondering if this was expected behavior with strict loading? It feels like a bug to me.

File: activerecord/lib/active_record/fixtures.rb [hunk 2]
```diff
@@ -773,9 +773,15 @@ def [](key)
...
     def find
       raise FixtureClassNotFound, "No class attached to find." unless model_class
-      model_class.unscoped do
+      object = model_class.unscoped do
         model_class.find(fixture[model_class.primary_key])
       end
+      # When loading a fixture (eg. `animals(:cat)`) you can't provide `include`
+      # statements, which are generally needed to avoid strict loading violations.
+      # To work around this, we just disable strict loading checks on models loaded
+      # through a fixture method.
```
Comment (eileencodes) [inline activerecord/lib/active_record/fixtures.rb:782]: Can you shorten this comment to "Fixtures can't be eagerly loaded"

File: activerecord/test/cases/fixtures_test.rb [hunk 3]
```diff
@@ -1589,3 +1591,19 @@ def readonly_config
...
+class StrictLoadingFixturesTest < ActiveRecord::TestCase
+  fixtures :strict_zines
+
+  test "strict loading violations are ignored on fixtures" do
+    ActiveRecord::FixtureSet.reset_cache
+    create_fixtures("strict_zines")
+
+    strict_zines(:going_out).interests.to_a
+
+    # by constrast, loading the record without a fixture call raises
```
Comment (eileencodes) [inline activerecord/test/cases/fixtures_test.rb:1604]:
```suggestion
```

File: activerecord/test/cases/fixtures_test.rb [hunk 4]
```diff
@@ -1589,3 +1591,19 @@ def readonly_config
...
 end
+
+class StrictLoadingFixturesTest < ActiveRecord::TestCase
+  fixtures :strict_zines
+
+  test "strict loading violations are ignored on fixtures" do
+    ActiveRecord::FixtureSet.reset_cache
+    create_fixtures("strict_zines")
+
+    strict_zines(:going_out).interests.to_a
```
Comment (eileencodes) [inline activerecord/test/cases/fixtures_test.rb:1602]: Can you assert the contents of this array or remove it? Then you don't need the comment below.

File: activerecord/test/cases/fixtures_test.rb [hunk 5]
```diff
@@ -1589,3 +1591,19 @@ def readonly_config
       end
   end
 end
+
+class StrictLoadingFixturesTest < ActiveRecord::TestCase
+  fixtures :strict_zines
+
+  test "strict loading violations are ignored on fixtures" do
```
Comment (eileencodes) [inline activerecord/test/cases/fixtures_test.rb:1598]: Can you move this test to the strict loading test file?

**Review by eileencodes — CHANGES_REQUESTED** [review]
Thanks for the fix, I left some requested changes.

**Comment by ghiculescu** [comment]
Thanks @eileencodes - I made the requested changes.

**Comment by eileencodes** [comment]
Thank you @ghiculescu!

**Comment by ghiculescu** [comment]
Thanks @eileencodes - is there a way to get that fix into the next 6.1.x release?

**Comment by eileencodes** [comment]
Yes, thank you for the reminder. I backported it in ddec53519e2c61665761e26d8c91a18724570b96

**Comment by sedubois** [comment]
With this PR merged, once `strict_loading` is used, the danger is that there is no more parity between test and prod, as existing tests using fixtures will still run fine but `ActiveRecord::StrictLoadingViolationError` can be raised in prod. In a way, any test that uses fixtures can't really be trusted any more.

What is the testing strategy to be able to use `strict_loading`? Do we need to rewrite all our tests to remove fixtures?

**Comment by ghiculescu** [comment]
@sedubois this PR should only remove prevent `ActiveRecord::StrictLoadingViolationError` on instances of fixtures you access directly, eg `strict_zines(:going_out)`. If you access objects through an AR query (`StrictZine.first`) strict loading still applies. And if you write controller/integration tests, any strict loading violations in the app should still raise.

Do you have an example of a test that's problematic with this change?

**Comment by sedubois** [comment]
@ghiculescu thanks very much. I probably misunderstood as I can't reproduce right now, otherwise I will get back with an example.

**Pull request merged**